chalk-solve/src/clauses.rs

```diff
@@ -169,10 +169,31 @@ fn program_clauses_that_could_match(
                 }
             }
 
-            // If the self type is `dyn Foo` (or `impl Foo`), then we generate clauses like:
+            // If the self type `S` is a `dyn trait` type, we wish to generate program-clauses
+            // that indicates that it implements its own traits. For example, a `dyn Write` type
+            // implements `Write` and so on.
+            //
+            // To see how this works, consider as an example the type `dyn Fn(&u8)`. This is
+            // really shorthand for `dyn for<'a> Fn<(&'a u8), Output = ()>`, and we represent
+            // that type as something like this:
+            //
+            // ```
+            // dyn(exists<T> {
+            //     forall<'a> { Implemented(T: Fn<'a>) },
+            //     forall<'a> { ProjectionEq(<T as Fn<'a>>::Output, ()) },
+            // })
+            // ```
+            //
+            // so what we wish to do is to generate program clauses of the form:
+            //
+            // ```
+            // forall<'a> { Implemented(dyn Fn(&u8): Fn<(&'a u8)>) }
+            // ```
+            //
+            // or
```

Comment:
nit: this “or” tripped me up at first, maybe add something about “, whatever the specified goal is trying to prove” (or don’t, it’s sorta obvious from context)

Reply:
improved